### PR TITLE
[IMP] disable mako compatibility in email templates

### LINKED ISSUE #45611: Disable line statements in jinja
Impacted versions: all

Steps to reproduce:

1. Use Odoo's HTML editor and edit any template that contains jinja line statements.

Current behavior:

The HTML editor does whatever it wants with whitespace, randomly rendering the template unusable.

Expected behavior:

Just changing one phrase in a template shouldn't break it.

To avoid that, the easiest way is to disable (and remove from Odoo source code) jinja line statements. This way, templates will be required to transform:

```jinja
% if condition:
    <p>True!</p>
% endif
```

into:

```jinja
<% if condition %>
    <p>True!</p>
<% endif %>
```

It's more verbose but it's unaffected by whitespaces.

Note: If you write `<% if condition %>` using the HTML editor, it converts it into `&lt;% if condition %&gt;`, making it impossible to workaround this problem using the UI, not even in developer mode. Possibly Odoo should switch to standard jinja tags `{% ... %}` to avoid that problem.

@Tecnativa TT22275

### PR DESCRIPTION
**Pull request by yajo**
- Blocks will now have to be in form `{% if condition %}` instead of old `% if condition:` or `<% if condition %>`.
- Comments are now `{# comment #}` instead of `<%doc> comment </%doc>`.
-  remove mako and inline jinja blocks

This will fix #45611.

@Tecnativa TT22275

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 1]
```diff
@@ -1975,21 +1975,21 @@ def test_00_sanitize(self):
...
         <td>
     </tr>
     <tr class="custom_class">
         This is some html.
     </tr>
-    % endif
+    {% endif %}
     <tr>
-%if object.dummy_field:
+{% if  object.dummy_field %}
```
Comment (MiquelRForgeFlow) [inline odoo/addons/test_new_api/tests/test_new_fields.py]:
```suggestion
{% if object.dummy_field %}
```

File: addons/stock_sms/data/sms_data.xml [hunk 2]
```diff
@@ -5,14 +5,14 @@
             <field name="name">Delivery: Send by SMS Text Message</field>
             <field name="model_id" ref="stock.model_stock_picking"/>
             <field name="body">
-                %if object.origin:
+                {% if  object.origin %}
```
Comment (MiquelRForgeFlow) [inline addons/stock_sms/data/sms_data.xml]:
```suggestion
                {% if object.origin %}
```

File: addons/stock/data/mail_template_data.xml [hunk 3]
```diff
@@ -10,23 +10,23 @@
     <p style="margin: 0px; padding: 0px; font-size: 13px;">
         Hello ${object.partner_id.name},<br/><br/>
         We are glad to inform you that your order has been shipped.
-        %if object.carrier_tracking_ref:
+        {% if  object.carrier_tracking_ref %}
             Your tracking reference is
             <strong>
-            %if object.carrier_tracking_url:
+            {% if  object.carrier_tracking_url %}
```
Comment (MiquelRForgeFlow) [inline addons/stock/data/mail_template_data.xml]:
```suggestion
            {% if object.carrier_tracking_url %}
```

File: addons/stock/data/mail_template_data.xml [hunk 4]
```diff
@@ -10,23 +10,23 @@
     <p style="margin: 0px; padding: 0px; font-size: 13px;">
         Hello ${object.partner_id.name},<br/><br/>
         We are glad to inform you that your order has been shipped.
-        %if object.carrier_tracking_ref:
+        {% if  object.carrier_tracking_ref %}
```
Comment (MiquelRForgeFlow) [inline addons/stock/data/mail_template_data.xml]:
```suggestion
        {% if object.carrier_tracking_ref %}
```

File: addons/digest/data/digest_template_data.xml [hunk 5]
```diff
@@ -32,16 +32,16 @@
...
                 </td></tr>
             </table>
-            % for kpi in kpis:
+            {% for kpi in kpis %}
                 <table style="border-spacing: 0; width: 100%; max-width: 600px;">
                     <tr>
                         <td style="border-collapse: collapse; background-color: #ffffff; border-left: 1px solid #e7e7e7; border-right: 1px solid #e7e7e7; line-height: 21px; padding: 0 20px 10px 20px; text-align: left;"><br/>
                             <span style="color: #3d466e; font-size: 18px; font-weight: 500; line-height: 23px;">${object.fields_get()[kpi]['string']}</span>
-                            %if kpi in kpi_actions:
+                            {% if  kpi in kpi_actions %}
```
Comment (MiquelRForgeFlow) [inline addons/digest/data/digest_template_data.xml]:
```suggestion
                            {% if kpi in kpi_actions %}
```

**Comment by Yajo** [comment]
@tde-banana-odoo @mart-e would you mind considering this ASAP? It seems the kind of PR that will keep on getting conflicts and will be left for later and get more conflicts and won't land until v15 and this is a  real problem. Quoting from a customer today:

> I think that the functionality of template edition is 100% unusable for end users. It's not normal that, when editing plain text -- just 1 phrase -- the whole template gets broken

**Comment by tde-banana-odoo** [comment]
Hello,

Thanks for the proposal. Indeed it should help to avoid getting a mess due to &gt and &lt within html.

I would say that my main discussion points are

  * retro compatibility: as templates could be in no update, people could have templates still using the old syntax. Or they could simply create new templates with the old syntax because they don't know it changed. What do you think about that ? It seems there is no way to allow both new and old syntax. Maybe we could do a pre-processing before compiling the template to replace old mako to new jinja in the string itself. At least grep for "% if, % endif, % for, % endfor, % set".
  * an alternative would be to check for "<doc", "<%", ... and try to do the "old rendering" on it;
  * migration: will have to check if it is doable to automatically migrate templates
  * enterprise templates should be updated as well;

Anyway, I think this should be validated by our testing team, meaning I create an internal task based on this PR in addition to discussion done here.

cc @odony

**Comment by Yajo** [comment]
> * retro compatibility: as templates could be in no update, people could have templates still using the old syntax. [...] What do you think about that ? It seems there is no way to allow both new and old syntax. Maybe we could do a pre-processing before compiling the template to replace old mako to new jinja in the string itself. At least grep for "% if, % endif, % for, % endfor, % set".

Migration scripts should handle that.

> Or they could simply create new templates with the old syntax because they don't know it changed.

Document it properly.

The fact of a user creating a old-syntax template from scratch is very unlikely because:

- It was broken before, unless using developer mode... and knowing HTML. :man_shrugging: 
- ... but if you're one of those guys, you also have still the preview feature in UI also.

Notice that I didn't replace `${ ... }` for jinja default's `{{ ... }}` because I understand that's the most used feature and it doesn't produce any of the errors mentioned in #45611.

> * an alternative would be to check for "<doc", "<%", ... and try to do the "old rendering" on it;

If your migration scripts are good enough, none of those should exist.

> * migration: will have to check if it is doable to automatically migrate templates

Yes, this is required of course. And it is doable for sure :wink:

**Comment by Yajo** [comment]
> It seems the kind of PR that will keep on getting conflicts

The prophecy started its fulfillment... :roll_eyes: 

Please tell me when you're gonna merge so I just do the conflict resolution once.

Thanks!

**Comment by Yajo** [comment]
I close this because keeping track of all templates on master is not something feasible for somebody from the community. You can base your definitive fix on this PR if you want. The list of conflicts will only grow as time goes by. Let's focus the discussion on #45611.

**Pull request closed without merging**